ga4gh/versioning.py

```diff
@@ -0,0 +1,152 @@
+Each RouteGroup specifies a set of rules which, when satisfied,
+users are allowed to access its routes.  These rules are:
+    1) a lifecycle
+    2) enabled and disabled versions
+
+Right now, three lifecycle stages are defined:
+    development - the default; endpoints that are under active development;
+        only can be accessed in development mode
+    active - live and stable; able to be accessed in any mode
+    deactivated - not able to be accessed in any mode
+
+Enabled and disabled versions:
+    these two versions specify the min and max versions at which the
+    endpoints can be accessed
+"""
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import functools
+
+
+class Lifecycle(object):
+    """
+    A lifecycle stage for a route group
+    """
+    # TODO more lifecycle stages?
+    # having a deprecated status doesn't really make any sense if there's
+    # no programatic difference from ACTIVE (for instance, an endpointStatus
+    # field in the response json indicating deprecation)
+    DEVELOPMENT = "DEVELOPMENT"
+    ACTIVE = "ACTIVE"
+    DEACTIVATED = "DEACTIVATED"
+
+
+class Version(object):
+    """
+    A major/minor/revision version tag
+    """
+    @classmethod
+    def parseString(cls, versionString):
+        versions = versionString.strip('vV').split('.')
+        return Version(*versions)
+
+    def __init__(self, major, minor, revision):
+        self.major = major
+        self.minor = minor
+        self.revision = revision
+
+    def __cmp__(self, other):
+        majorCmp = cmp(self.major, other.major)
+        if majorCmp != 0:
+            return majorCmp
+        minorCmp = cmp(self.minor, other.minor)
+        if minorCmp != 0:
+            return minorCmp
+        else:
+            return cmp(self.revision, other.revision)
+
+    def __hash__(self):
+        return hash((self.major, self.minor, self.revision))
+
+    def __eq__(self, other):
+        return (self.major == other.major and
+                self.minor == other.minor and
+                self.revision == other.revision)
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+
+class RouteGroup(object):
+    """
+    A group that a route belongs to
+    """
+    def __init__(self, lifecycleStage=Lifecycle.DEVELOPMENT,
+                 enabledVersion=None, disabledVersion=None):
+        self.lifecycleStage = lifecycleStage
+        if enabledVersion is not None:
+            self.enabledVersion = Version.parseString(enabledVersion)
+        else:
+            self.enabledVersion = None
+        if disabledVersion is not None:
+            self.disabledVersion = Version.parseString(disabledVersion)
+        else:
+            self.disabledVersion = None
+
+
+class VersioningManager(object):
+
+    # TODO externalize data fields to yaml file?
+    supportedVersionsData = ['0.5.1']
+
+    groupData = {
+        'V_0.5.1_RELEASE': {
```

Comment:
I *think* I can intuit how you intend this to be used, but could we have some examples on the bug (if not in commit message or comments) of how you'd meaninfully have multiple of these or how they'd change over time?